### PR TITLE
[None][perf] Allocate DSA indexer k-cache only for layers that own an indexer

### PR DESCRIPTION
**Pull request by Tabrizian**
## Dev Engineer Review

- Implements per-layer masked Indexer-K cache pool allocation for cross-layer sharing:
  - Builds Indexer-K pool rows only for “full indexer”/indexer-owning layers; shared layers are masked out.
  - Adds a compact global layer → masked pool-row mapping (`getIndexerKCachePoolLayerIdx`), and enforces rejection of masked-out layers/buffers on the consumer side (C++ and Python).
- Threads indexer-layer masking through cache/transceiver plumbing:
  - Adds `indexerKCacheLayerMask`/`indexerLayerNumPerPP` through KV cache manager construction, `CacheTransceiver`, and `kv_cache::CacheState` (including equality + `toString()` updates).
  - Updates Indexer-K specific cache transmission topology/routing (`targetIRanksForIndexerKCache`) and buffer-kind offset-ratio handling for `BufferKind::kKV_INDEXER`.
  - Updates PP resharding/split-concat logic to operate in “indexer layer space” using indexer-layer counts.
  - Adjusts Indexer-K cache transfer buffer sizing to derive sizes directly from the Indexer-K pool (avoiding attention-window iteration when Indexer-K transfer is enabled).
- Updates serialization and language bindings:
  - Extends `kv_cache::CacheState` serialization/deserialization with `indexerLayerNumPerPP`.
  - Nanobind Python bindings add `indexer_k_cache_layer_mask` and expose layer→pool-row mapping with masking enforcement.
  - Python NIXL/DSA path enforces ownership: shared (non-full) indexer layers have `None` pools and access is blocked.
- GLM 5.2 correctness/capacity:
  - Reduces Indexer-K cache accounting per token (55,224 → 47,700 bytes/token), increasing effective KV capacity by ~15.8%.
  - Dense layouts remain unchanged when cross-layer sharing is not used; shared layers do not allocate Indexer-K pools; model outputs are intended to remain bitwise-identical.
- Disaggregated serving / resource-extractor behavior:
  - Adds fail-fast validation in the disaggregated KV extractor for unsupported masked ownership patterns, with guidance to disable disaggregated serving for cross-layer indexer sharing.

### Test-code changes (files outside `tests/integration/test_lists/`)
- `cpp/tests/unit_tests/batch_manager/kvCacheManagerTest.cpp`
  - Added/updated unit coverage for masked Indexer-K pool row mapping and dense default behavior.
- `cpp/tests/unit_tests/executor/serializeUtilsTest.cpp`
  - Extended `CacheStateIndexerKCache` serialization/deserialization round-trip to include `indexerKCacheUseFp4` and `indexerLayerNumPerPP`.
- `cpp/tests/unit_tests/multi_gpu/cacheTransceiverTest.cpp`
  - Added `AsymmetricalCacheTestWithIndexerLayerMask` and updated Indexer-K block-data generation/verification to use indexer-layer counts.

**Coverage in `tests/integration/test_lists/`:** these are unit-test binary changes (not directly represented in the integration test list files).  
**Verdict (unit tests):** sufficient.

### Test-list / integration changes (files under `tests/integration/test_lists/`)
- `tests/integration/defs/accuracy/test_disaggregated_serving.py`
  - Renamed `TestGLM52NVFP4.test_nvfp4_nixl_python[...]` → `TestGLM52NVFP4.test_nvfp4_nixl[...]`.
- `tests/integration/test_lists/qa/llm_function_core.txt`
  - Updated the GLM52 NVFP4 NIXL QA entry for `cache_mgr_v1` to point to `test_nvfp4_nixl[...]` instead of `test_nvfp4_nixl_python[...]`.
- `tests/integration/test_lists/test-db/l0_dgx_b200.yml`
  - Updated the GLM52 NVFP4 coverage for `cache_mgr_v1` similarly to use `test_nvfp4_nixl[...]` (timeout unchanged).
- `tests/integration/test_lists/waives.txt`
  - Added a skipped multi-GPU test entry for `cpp/test_multi_gpu.py::test_cache_transceiver` with `8proc-ucx_kvcache-90`, referencing NVBUG 5838199.

**Verdict (integration lists):** needs follow-up (CI failures were reported for specific helper commit runs in the PR context; additionally, CBTS coverage sufficiency for the updated list entries is not confirmed here).

### Overall QA verdict
**needs follow-up** due to the reported L0 pipeline failures and to validate end-to-end stability across the affected masked Indexer-K transceiver/PP-resharding and Python NIXL paths.

## Description

GLM 5.2 (`glm_moe_dsa`, served via the DeepSeek-V3 path) uses DeepSeek Sparse Attention with **cross-layer indexer sharing**: only 21 of 78 layers own an indexer ("full" layers); the remaining 57 "shared" layers reuse the previous full layer's top-k indices and never touch the indexer k-cache. The V1 `KVCacheManager` allocated the indexer k-cache stride for **every** layer, so the shared layers' indexer pools were dead memory — 13.6% of the whole KV cache footprint (55,224 B/token instead of 47,700 B/token, i.e. **+15.8% effective KV token capacity per rank** once removed). On long-context agentic workloads with block reuse, KV capacity directly drives cache hit rate and prefill throughput.

**Commit 1 — per-layer masked indexer k-cache pool (KVCacheManager V1):**
- New optional per-local-layer `indexerKCacheLayerMask` through the `KVCacheManager`/`BlockManager`/`WindowBlockManager` ctors; `createIndexerKCachePools` now allocates one pool row per masked-in layer (rows follow the KV pool's layer order) and records the layer → pool-row map (`getIndexerKCachePoolLayerIdx`).
- nanobind: `indexer_k_cache_layer_mask` ctor kwarg; `get_indexer_k_cache_pool_data` translates local layer → pool row and rejects masked-out layers.
- `DSACacheManager` derives the mask from `to_sparse_params(pretrained_config, layer_idx).is_full_indexer_layer` — the exact source of truth MLA uses to decide whether a layer constructs an `Indexer` module (trailing spec/MTP layers always resolve to full). Shared layers get `None` pool entries and assert if asked for indexer buffers.
- Both memory estimators (`get_cache_size_per_token` / `get_cache_bytes_per_token`) charge indexer bytes for full-indexer layers only, PP-sliced consistently with the pool.
- Models without cross-layer sharing (e.g. DeepSeek V3.2) resolve to an all-full mask and keep the dense legacy layout bit-for-bit. DeepSeek-V4 uses its own cache manager and is unaffected. `kv_cache_manager_v2.py` is untouched.
- The Python KV-transceiver runtime (NIXL native path) rejects masked pools with a clear `NotImplementedError` (its FLAT pool-view contract assumes one row per layer); `TestGLM52NVFP4::test_nvfp4_nixl_python` is switched to the C++ NIXL transceiver and renamed `test_nvfp4_nixl` (CI list entries updated).

**Commit 2 — C++ cache transceiver support for the masked pool:**
- `CacheState` carries per-PP indexer layer counts (`indexerLayerNumPerPP`; empty = dense fallback to the attention counts), wired through the ctors, `operator==`, `toString`, and serialization.
- New `targetIRanksForIndexerKCache` keeps the attention pass's rank topology (same connection set) but recomputes per-peer layer counts in indexer layer space (interval intersections of the per-PP indexer prefix sums). `splitKVCache`/`concatKVCache` use it — plus the self indexer count as the kernel layer count — for the indexer pass.
- `MLACacheFormatter` sizes indexer-pass buffers with the indexer counts; `inquireSupport` validates indexer layout equality and equal totals, and rejects PP partitions that would produce zero-sized indexer exchanges (a rank or layer overlap without a full-indexer layer).
- The NIXL/Mooncake agent path computes a dedicated indexer-pass write-offset ratio in indexer layer space (it previously shared the attention-space KV ratio, which would corrupt masked-pool transfers under PP resharding).
- `CacheTransceiver` takes `indexerLayerNumPerPP`, validates it against the local pool row count; the Python transceiver gathers it over PP from the manager's indexer layer mask. `CacheTransBufferManager` sizes the indexer transfer buffer from the pool's actual (masked) layer count.

Expected GLM 5.2 server-start accounting: `kv size per token is 47700 bytes/token` (was 55,224) and ~+15.8% `max tokens in paged KV cache` at the same `kv_cache_free_gpu_memory_fraction`. Shared layers never used the pools, so model outputs are bitwise-identical.

## Test Coverage

New tests (all passing on H100):
- `cpp/tests/unit_tests/batch_manager/kvCacheManagerTest.cpp` — `IndexerKCachePoolLayerMaskTest` (masked pool shape, row mapping, block size) and `IndexerKCachePoolDenseDefaultTest` (dense layout unchanged without a mask).
- `cpp/tests/unit_tests/executor/serializeUtilsTest.cpp` — `CacheState` round-trip including `indexerLayerNumPerPP`.
- `cpp/tests/unit_tests/multi_gpu/cacheTransceiverTest.cpp` — a new `AsymmetricalCacheTestWithIndexerLayerMask` end-to-end masked-indexer transfer with ctx/gen PP resharding (also fixes a positional off-by-one in the existing `CacheState` construction where `isIndexerKCache` landed on `enablePartialReuse`).

Regression: full `kvCacheManagerTest` and `serializeUtilsTest` suites pass; existing DSA indexer scatter/roundtrip GPU tests pass (dense pool → identity mapping). End-to-end guard: `TestGLM52::test_nvfp4` (GSM8K) exercises the masked pool implicitly; `TestDeepSeekV32` covers the dense path.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by BowenFu — APPROVED** [review]
LGTM — fully gated to the DSA indexer K-cache (enableIndexerKCache); the indexerKCacheLayerMask defaults to nullopt → dense allocation byte-identical to today, and is_full_indexer_layer defaults True so DeepSeek-V3.2 still allocates every layer. Non-DSA KV-cache allocation is untouched, and under-allocation fails loudly (mask-size assert, -1 pool idx for masked layers, non-None pool checks, transceiver rejects zero-indexer-row ranks). Only cross-layer-sharing models shrink.

**Review by StanleySun639 — APPROVED** [review]

**Review by eopXD — APPROVED** [review]
Looks good to me

**Review by Shixiaowei02 — APPROVED** [review]
Thank you for the update! As a heads-up, @chuangz0 is working on enabling KV Manager v2 + Transceiver v2 by default and will also be making or contributing to the related indexer mask changes. Please make sure to stay aligned and keep each other informed on the corresponding updates.

**Comment by Tabrizian** [comment]
/bot run --disable-fail-fast --extra-stage "DGX_B200-8_GPUs-PyTorch-1, DGX_B200-8_GPUs-PyTorch-2, DGX_B200-8_GPUs-PyTorch-3, DGX_B200-8_GPUs-PyTorch-4"

**Review by chienchunhung — APPROVED** [review]
Thanks for the PR!

**Review by cascade812 — APPROVED** [review]

File: tensorrt_llm/_torch/disaggregation/resource/kv_extractor.py [hunk 1]
```diff
@@ -199,9 +199,31 @@ def build_page_table(kv_cache_manager: KVCacheManager) -> KVCachePageTable:
 
         # Indexer K cache support
         if getattr(kv_cache_manager, "enable_indexer_k_cache", False):
+            local_indexer_mask = getattr(kv_cache_manager, "indexer_k_cache_local_layer_mask", None)
```
Comment (chuangz0) [inline tensorrt_llm/_torch/disaggregation/resource/kv_extractor.py]: Since the PR https://github.com/NVIDIA/TensorRT-LLM/pull/16017 has been merged, the basic work is enough for indexer pool. Aside from the changes we need to make to the build_page_table,  I believe we’ll only need to make a few changes to support `indexer_mask`. Perhaps this PR can also implement it, or it could be implemented in a subsequent PR.

Comment (Tabrizian) [reply tensorrt_llm/_torch/disaggregation/resource/kv_extractor.py]: I'll do this in a follow up PR.

**Review by chuangz0 — APPROVED** [review]

**Review by tburt-nv — APPROVED** [review]

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 2]
```diff
@@ -3366,27 +3389,40 @@ def __init__(
...
+        local_mask = self.indexer_k_cache_local_layer_mask
         self.indexer_k_cache_pool_per_layer = [
-            self.get_indexer_k_cache_pool_data(layer_idx)
-            for layer_idx in range(self.num_local_layers)
+            self.get_indexer_k_cache_pool_data(local_offset)
+            if local_mask[local_offset] else None
+            for local_offset in range(self.num_local_layers)
         ]
+        num_full = sum(local_mask)
+        if num_full < self.num_local_layers:
```
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py:3410]: Just curious, when will this happen?

Comment (Tabrizian) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa.py:3410]: Every 3 or 4 layers of GLM5.2 own an Indexer K-Cache.

**Comment by Tabrizian** [comment]
@fredricz-20070104 I was able to reproduce the UCX failure on main too.
